pandas/core/util/hashing.py

```diff
@@ -139,25 +167,25 @@ def hash_pandas_object(
     return h
 
 
-def hash_tuples(vals, encoding="utf8", hash_key: str = _default_hash_key):
+def hash_tuples(
+    vals: Union[MultiIndex, List[Tuple]],
```

Comment:
whats the rule for Sequence vs Iterable?

In this case, we only get here with MultiIndex and List[Tuple].  I get the wider-is-better for public APIs, but for purely internal code strictness seems beneficial.  is that distinction not relevant?

Reply:
AFAIK mypy works by checking each function is isolation. i.e. it checks that all the code within the function is compatible with the function annotation. i.e. checks the types in function calls within the body with the type parameters of the called functions.

so doing the reverse of this, i.e. annotating with the actual types passed seems like extra work imo. (and could be simply achieved using MonkeyType)

Iterable because `c_is_list_like` accepts an iterable as list-like

Using List is bad practice. function code could modify the contents of the List. If the parameter is typed as an immutable container, mypy won't allow it.

Reply:
thanks, updated